### DISCUSSION TITLE
3D Box help connecting edges

### DISCUSSION THREAD

**Discussion by mathiasgrimm** [ImageMagick/ImageMagick#4491] [Help] [answered]
Hey guys, I need some help, php specific (or not).
I have a painting and a canvas which is in perspective. I need to place the image on top of the canvas, however, to do that I created 2 images, first one is the front and the second it's the side.

here are the images and the current result.

and the result:

The code is:

```php
private function mockup($imagePath, $code)
    {
        $framePath = __DIR__ . "/resource/images/retangulo/{$code}/frame.jpg";

        $frame = new \imagick($framePath);
        $frame->setColorspace(imagick::COLORSPACE_SRGB);
        $frame->setImageFormat('png');
        $frame->setImageVirtualPixelMethod(\Imagick::VIRTUALPIXELMETHOD_TRANSPARENT);
        $frame->setImageMatte(true);

        $imageOriginal = new \imagick($imagePath);
        $imageOriginal->setColorspace(imagick::COLORSPACE_SRGB);
        $imageOriginal->setImageFormat('png');
        $imageOriginal->setImageVirtualPixelMethod(\Imagick::VIRTUALPIXELMETHOD_TRANSPARENT);
        $imageOriginal->setImageMatte(true);

        $sidePx = 50;
        $front = clone $imageOriginal;
        $front->cropImage(
            $front->getImageWidth() - ($sidePx * 2),
            $front->getImageHeight() - ($sidePx * 2),
            $sidePx,
            $sidePx,
        );

        $front->resizeimage(
            $frame->getImageWidth(),
            $frame->getImageHeight(),
            \Imagick::FILTER_LANCZOS,
            1
        );

        $controlPoints = [
            0, 0, 241, 139, // top left
            0, $frame->getImageHeight(), 241, 823, // bottom left
            $frame->getImageWidth(), $frame->getImageHeight(), 578, 860, // bottom right
            $frame->getImageWidth(), 0, 580, 104, // top right
        ];

        $front->distortImage(\Imagick::DISTORTION_PERSPECTIVE, $controlPoints, false);

        $rightSide = clone $imageOriginal;

        $rightSide->cropImage(
            $sidePx,
            $rightSide->getImageHeight() - (2 * $sidePx),
            $rightSide->getImageWidth() - $sidePx,
            $sidePx,
        );

        $rightSide->resizeimage(
            $frame->getImageWidth(),
            $frame->getImageHeight(),
            \Imagick::FILTER_LANCZOS,
            1
        );

        $controlPoints = [
            0, 0, 580, 104, // top left
            0, $frame->getImageHeight(), 578, 860, // bottom left
            $frame->getImageWidth(), $frame->getImageHeight(), 605, 846, // bottom right
            $frame->getImageWidth(), 0, 605, 108, // top right
        ];

        $rightSide->distortImage(\Imagick::DISTORTION_PERSPECTIVE, $controlPoints, false);
        $rightSide->compositeImage($front, \Imagick::COMPOSITE_PLUS, 0, 0);

         $frame->compositeImage($rightSide, \Imagick::COMPOSITE_OVER, 0, 0);

        return $frame;
    }
```

**Comment by fmw42** [comment]
What version of ImageMagick and what platform/OS?  If on Unix-like system, see my 3Dbox shell script running ImageMagick. It can be called from PHP exec().  Also see https://github.com/ImageMagick/ImageMagick/discussions/4231  and https://legacy.imagemagick.org/Usage/distorts/#box3d

**Comment by snibgo** [comment]
@mathiasgrimm: You show one input, but not frame.jpg.

You show code and an output. And .. what? Perhaps you don't like something about the output. What don't you like?

Your result doesn't simulate a 3D box in perspective. A simulation would have real-world vertical lines either parallel or converging to one point. It would have the horizontal lines of both panels converging to points that were on a horizontal line (the "horizon" line).

**Comment by mathiasgrimm** [comment]
Hi @fmw42 , I've tried to do that of box3d doc but wasn't able to fully translate it to php. I don't want to use the exec function unless I can't really do what im trying to do with the native methods.

@snibgo Hi,
Yep, the file names in the function are not the same as the attached here but its easy to know which one is which.
From that 3DBox example, looks like its possible to connect the 2 images/layers so that the all the edges would be connected.

I am failing real badly to make this happen. 
First question would be:
1 - Is it possible to make ALL the edges connect 100% so that it looks like a 3d image?
2 - if its possible, is it possible via php (without using exec or other system call)?

**Comment by mathiasgrimm** [comment]
Btw, thanks guys for replying. Seems like this is the only place I can get help.

**Comment by mathiasgrimm** [comment]
I am using Ubuntu 20
Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

And PHP 8.0

**Comment by snibgo** [comment]
>  1 - Is it possible to make ALL the edges connect 100% so that it looks like a 3d image?

Yes, it is possible. But you (or the script) need to understand perspective. For example, you move the front panel corners to (241,139), (241,823), (578,860) and (580,104). Where did those numbers come from? Did you calculate them or just guess? The left side is vertical at x=241, but the right side slopes inwards from x=580 at the top to x=578 at the bottom. That is fair enough, but then the right side of the right panel should also slope inwards. 

>  2 - if its possible, is it possible via php (without using exec or other system call)?

I don't use IMagick and PHP, but I expect it is possible.

**Reply by mathiasgrimm** [reply]
Awesome! will check what you said and get back.

All x and y are calculated based on the frame. I resize the frame, and the 2 parts of the painting with the same size so that I have a common denominator.
I get the X,Y of the front and place them on the edges of the frame (the red part), mind you that the frame image is actually bigger than the red part, is has some white area around it.
Then, next, I take X,Y for the side and place them also on the edges of the side.

What are the premisses for connecting the edges? Do the 2 images (front and side (spine)) need to be different layers and then be merged? Is there a specific order I need to run the commands? For example, In my code, I distort separately the side and the front and then merge them together using the composite with COMPOSITE_PLUS, like in the 3DBox example ,then I merge them to the frame.
But even when I take the frame out of the questions and work only with the front and side, I can get them edge-connected.

So far thanks for the help!

**Reply by mathiasgrimm** [reply]
also, do I actually need to split the painting in two? that is what happens on the 3DBox example but I wonder if I could keep it as a single Image (to avoid layers/merges) and instead have more control points on a single distort command

**Comment by mathiasgrimm** [comment, marked as answer]
I've done some tests now and just changing the distort from PERSPECTIVE to BILINEAR made it work
